File: typescript/statisticsCalculator.ts
export class StatisticsCalculator {
  static calculateMean(numbers: number[]): number {
    if (numbers.length === 0) {
      throw new Error("Empty array");
    }
    
    const sum = numbers.reduce((acc, val) => acc + val, 0);
    return sum / numbers.length;
  }
  
  static calculateMedian(numbers: number[]): number {
    if (numbers.length === 0) {
      throw new Error("Empty array");
    }
    
    const sortedNumbers = [...numbers].sort((a, b) => a - b);
    
    if (sortedNumbers.length % 2 === 0) {
      const midIndex = sortedNumbers.length / 2;
      return (sortedNumbers[midIndex - 1] + sortedNumbers[midIndex]) / 2;
    } else {
      return sortedNumbers[Math.floor(sortedNumbers.length / 2)];
    }
  }
  
  static calculateVariance(numbers: number[]): number {
    if (numbers.length === 0) {
      throw new Error("Empty array");
    }
    
    const mean = this.calculateMean(numbers);
    const sumOfSquaredDifferences = numbers.reduce((acc, val) => {
      return acc + Math.pow(val - mean, 2);
    }, 0);
    
    return sumOfSquaredDifferences / numbers.length;
  }
  
  static calculateStandardDeviation(numbers: number[]): number {
    return Math.sqrt(this.calculateVariance(numbers));
  }
  
  static calculateMode(numbers: number[]): number[] {
    if (numbers.length === 0) {
      throw new Error("Empty array");
    }
    
    const frequencyMap: Record<number, number> = {};
    let maxFrequency = 0;
    let modes: number[] = [];
    
    numbers.forEach(num => {
      frequencyMap[num] = (frequencyMap[num] || 0) + 1;
      
      if (frequencyMap[num] > maxFrequency) {
        maxFrequency = frequencyMap[num];
        modes = [num];
      } else if (frequencyMap[num] === maxFrequency) {
        modes.push(num);
      }
    });
    
    return modes;
  }
}

File: java/StatisticsCalculator.java
package java;

import java.util.Arrays;

public class StatisticsCalculator {
    public static double calculateMean(int[] numbers) {
        if (numbers.length == 0) {
            throw new IllegalArgumentException("Empty array");
        }
        
        int sum = 0;
        for (int number : numbers) {
            sum += number;
        }
        
        return (double) sum / numbers.length;
    }
    
    public static double calculateMedian(int[] numbers) {
        if (numbers.length == 0) {
            throw new IllegalArgumentException("Empty array");
        }
        
        int[] sortedNumbers = Arrays.copyOf(numbers, numbers.length);
        Arrays.sort(sortedNumbers);
        
        if (sortedNumbers.length % 2 == 0) {
            int midIndex = sortedNumbers.length / 2;
            return (sortedNumbers[midIndex - 1] + sortedNumbers[midIndex]) / 2.0;
        } else {
            return sortedNumbers[sortedNumbers.length / 2];
        }
    }
    
    public static double calculateVariance(int[] numbers) {
        if (numbers.length == 0) {
            throw new IllegalArgumentException("Empty array");
        }
        
        double mean = calculateMean(numbers);
        double sumOfSquaredDifferences = 0;
        
        for (int number : numbers) {
            sumOfSquaredDifferences += Math.pow(number - mean, 2);
        }
        
        return sumOfSquaredDifferences / numbers.length;
    }
    
    public static double calculateStandardDeviation(int[] numbers) {
        return Math.sqrt(calculateVariance(numbers));
    }
}